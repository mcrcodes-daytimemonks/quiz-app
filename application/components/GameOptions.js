import { useRouter } from "next/router";
import { useEffect, useState } from "react";
import css from "../styles/GameOptions.module.css";

const INIT = {
  QUESTION_LIMIT: {
    MIN: "1",
    MAX: "5",
    DEFAULT: "3",
  },
  CATEGORIES: ["JavaScript", "HTML"],
};

const GameOptions = () => {
  const router = useRouter();
  const [selectedCategory, setSelectedCategory] = useState("");
  const [selectedQuestionLimit, setSelectedQuestionLimit] = useState(
    INIT.QUESTION_LIMIT.DEFAULT
  );

  const updateCategory = ({ target: { value: category } }) => {
    localStorage.setItem("category", category);
    setSelectedCategory(category);
  };

  const updateQuestionLimit = ({ target: { value: questionLimit } }) => {
    localStorage.setItem("questionLimit", questionLimit);
    setSelectedQuestionLimit(questionLimit);
  };

  const goToGame = (event) => {
    event.preventDefault();
    router.push("/game");
  };

  useEffect(() => {
    localStorage.setItem("questionLimit", INIT.QUESTION_LIMIT.DEFAULT);
  }, []);

  return (
    <form
      className={css.GameOptions__grid}
      name="game-options"
      onSubmit={goToGame}
    >
      <fieldset>
        <legend>Please choose a category</legend>

        {INIT.CATEGORIES.map((category) => (
          <div key={category}>
            <input
              checked={category === selectedCategory}
              id={category}
              name="category"
              onChange={updateCategory}
              required
              type="radio"
              value={category}
            />
            <label htmlFor={category}>{category}</label>
          </div>
        ))}
      </fieldset>

      <div>
        <label htmlFor="questionCount">Number of questions: </label>
        <input
          id="questionCount"
          max={INIT.QUESTION_LIMIT.MAX}
          min={INIT.QUESTION_LIMIT.MIN}
          onChange={updateQuestionLimit}
          required
          step="1"
          type="number"
          value={selectedQuestionLimit}
        />
      </div>
      <button type="submit">Play Game</button>
    </form>
  );
};

export default GameOptions;
