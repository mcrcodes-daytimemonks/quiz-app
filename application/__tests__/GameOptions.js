import { fireEvent, render, screen, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import GameOptions from "../components/GameOptions";

const INIT = {
  QUESTION_LIMIT: {
    MIN: "1",
    MAX: "5",
    DEFAULT: "3",
  },
  CATEGORIES: ["JavaScript", "HTML"],
};

const renderGameOptions = () => render(<GameOptions />);

describe("GameOptions", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderGameOptions();

    expect(asFragment()).toMatchSnapshot();
  });

  describe("Inputs form", () => {
    it("renders a form element to group the user inputs", () => {
      renderGameOptions();

      expect(screen.getByRole("form")).toBeInTheDocument();
    });
  });

  describe("Question Category input", () => {
    it("renders input/s to allow the user to select a question category", () => {
      renderGameOptions();
      const radioGroup = screen.getByRole("group", {
        name: /please choose a category/i,
      });

      expect(radioGroup).toBeInTheDocument();

      INIT.CATEGORIES.forEach((category) => {
        expect(
          within(radioGroup).getByRole("radio", { name: category })
        ).toBeInTheDocument();
      });
    });

    it("updates application state when the user selects a question category", () => {
      renderGameOptions();

      expect(localStorage.getItem("category")).toBeFalsy();

      const radioButtons = within(
        screen.getByRole("group", { name: /please choose a category/i })
      ).getAllByRole("radio");

      radioButtons.forEach((radioButton) => {
        fireEvent.click(radioButton);

        expect(localStorage.getItem("category")).toEqual(radioButton.value);
        expect(radioButton).toBeChecked();
      });
    });
  });

  describe("Number of Questions input", () => {
    it("renders a number input to allow the user to select the number of questions in a game", () => {
      renderGameOptions();

      expect(
        screen.getByRole("spinbutton", { name: /number of questions/i })
      ).toBeInTheDocument();
    });

    it("renders with a default value", () => {
      renderGameOptions();

      const numberInput = screen.getByRole("spinbutton", {
        name: /Number of questions:/i,
      });

      expect(numberInput.value).toEqual(INIT.QUESTION_LIMIT.DEFAULT);
    });

    it("updates application state when the user sets the number of questions", () => {
      renderGameOptions();
      const numberInput = screen.getByRole("spinbutton", {
        name: /Number of questions:/i,
      });

      expect(numberInput.value).toEqual(INIT.QUESTION_LIMIT.DEFAULT);
      expect(localStorage.getItem("questionLimit")).toEqual(
        INIT.QUESTION_LIMIT.DEFAULT
      );

      fireEvent.change(numberInput, {
        target: {
          value: Number(INIT.QUESTION_LIMIT.DEFAULT) + 1,
        },
      });

      expect(numberInput.value).toEqual(
        `${Number(INIT.QUESTION_LIMIT.DEFAULT) + 1}`
      );
      expect(localStorage.getItem("questionLimit")).toEqual(
        `${Number(INIT.QUESTION_LIMIT.DEFAULT) + 1}`
      );
    });
  });
});
