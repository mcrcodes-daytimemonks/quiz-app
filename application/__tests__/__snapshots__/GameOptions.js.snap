// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOptions matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="GameOptions__grid"
    name="game-options"
  >
    <fieldset>
      <legend>
        Please choose a category
      </legend>
      <div>
        <input
          id="JavaScript"
          name="category"
          required=""
          type="radio"
          value="JavaScript"
        />
        <label
          for="JavaScript"
        >
          JavaScript
        </label>
      </div>
      <div>
        <input
          id="HTML"
          name="category"
          required=""
          type="radio"
          value="HTML"
        />
        <label
          for="HTML"
        >
          HTML
        </label>
      </div>
    </fieldset>
    <div>
      <label
        for="questionCount"
      >
        Number of questions: 
      </label>
      <input
        id="questionCount"
        max="5"
        min="1"
        required=""
        step="1"
        type="number"
        value="3"
      />
    </div>
    <button
      type="submit"
    >
      Play Game
    </button>
  </form>
</DocumentFragment>
`;
